Use Morph's JSON API

chrismytton opened the pull request:
# What does this do?

This adds 2 new methods for getting data from a remote morph scraper, `MorphScraper::Database#data` and `MorphScraper::Database#query`. All of these methods return an array of hashes representing the records from the remote scraper's database.

``` ruby
db = MorphScraper::Database.new('tmtmtmtm/malta-parliament')

# Equivalent to 'SELECT * FROM data'
db.data

# Equivalent to 'SELECT * FROM terms'
db.data(:terms)

# Or just execute arbitrary queries
db.query('SELECT *, 5 as term FROM data')
```
# Why was this needed?

This gem already [has a method](https://github.com/everypolitician/morph_scraper-database/blob/bcfd6fb77d288d2e59ca5405fea9f561f1ba6db3/lib/morph_scraper/database.rb#L14-L21) for copying the database from another morph scraper into the current scraper's `data.sqlite`, but it completely overwrites what's in the current scrapers database. This is OK if you've just created a new scraper and want the new scraper's database to be identical to the old one, but falls down for anything more complicated than that.

To solve this problem this pull request adds the methods mentioned above to give you full programatic access to the contents of a morph scraper. This can be used by a scraper to import the partial contents of another scraper. It can also potentially be used as part of the everypolitician-data build process to make it easier to work with remote scrapers.
# Implementation notes

This is built on top of the [morph.io JSON API](https://morph.io/documentation/api), so most of the heavy lifting is done by them, with this gem just shuttling the data around and providing some convenience methods for accessing the data.
# Related issues

https://github.com/everypolitician/morph_scraper-database/pull/1 was my first attempt at this, but that adds extra dependencies and requires the user to have sqlite installed. With this version we just use http, so the requirements for consumers are kept to a minimum.

chrismytton committed:
Add section for replacing local database to README

This is so that I can add another section in the next commit which will
explain how to get a subset of the data from a scraper.

chrismytton committed:
Add details about getting data from a scraper to README

chrismytton committed:
Fix typo in README.md

loose -> lose

chrismytton committed:
Fix link to GitHub repo in README.md

chrismytton committed:
Move with_tmp_dir into test_helper.rb

This is a helper method, so it should go in the helper file rather than
clutter up the test files.

chrismytton committed:
Add stub_morph_query helper for testing

This makes it more obvious what's going on in the tests, rather than the
reader needing intimate knowledge of the API url structure.

chrismytton committed:
Add Database#data method

This allows you to get data from the default morph database, `data`.

chrismytton committed:
Allow data to take an argument specifying the table

This provides a simple shortcut for getting all of the data from a
specific scraper table.

chrismytton committed:
Add Database#query method

This allows you to run arbitrary SQL against a remote scraper.

chrismytton commented:
@tmtmtmtm 👀

chrismytton commented:
Note: if you've got a `MORPH_API_KEY` environment variable set then you can test this branch out by checking it out and using `bin/console`, which will load up the library. Then you can use the following code to see it in action:

``` ruby
MorphScraper::Database.new('everypolitician-scrapers/aland-lagting').data
```

tmtmtmtm reviewed (approved):
Yay. Much better.

Travis is failing, though.

chrismytton committed:
Ignore Metrics/BlockLength cop for tests

The tests have long blocks because we use Minitest::Spec which wraps all
of the tests in a file in a block, so turn this cop off for the test
directory.
